dags/common/france_travail/api.py

```diff
@@ -0,0 +1,199 @@
+        )
+
+        return rows
+
+    # We log which quarters have already been accessed by previous executions of this task
+    # If this cache is empty, we'll pull everything available from the API
+    logged_sessions_by_territory = json.loads(Variable.get("FT_INFORMATION_TERRITOIRE_PERIOD_LOG", "{}"))
+
+    # If a log is present, we make the assumptions that
+    # - the DAG has run successfully since the last quarter
+    # - the data we have for previous quarters don't need to be updated
+    # - the most recently updated quarter is the only one we are missing
+    log_key = f"{territory.type}_{territory.code}"
+    limit_to_most_recent_quarter = logged_sessions_by_territory and log_key in logged_sessions_by_territory
+
+    response = httpx.post(
+        url=f"{FT_JOBSEEKER_STATS_BASE_URL}/indicateur/stat-demandeurs",
+        headers={
+            "Accept": "application/json",
+            "Authorization": access_token,
+            "Content-Type": "application/json",
+        },
+        json={
+            "codeTypeTerritoire": territory.type,  # REG / DEP
+            "codeTerritoire": territory.code,
+            "codeTypeActivite": "CUMUL",  # CUMUL = All activities
+            "codeActivite": "CUMUL",
+            "codeTypePeriode": "TRIMESTRE",
+            "codeTypeNomenclature": "CATCAND",  # Stats sur le nombre des demandeurs d'emplois
+            "dernierePeriode": limit_to_most_recent_quarter,
+        },
+    )
+    # NOTE: the response JSON is a very large dictionary
+    response.raise_for_status()
+    response_data = response.json()
+    data = response_data.get("listeValeursParPeriode", [])
+
+    if not data:
+        logger.info("No data found for territory %s, skipping SQL import.", territory)
+        return
+
+    periods_returned = [datum["codePeriode"] for datum in data]
+    if limit_to_most_recent_quarter:
+        if set(periods_returned) - set(logged_sessions_by_territory[log_key]) == set():
+            logger.info("No new data for territory %s, skipping SQL import.", log_key)
+            return
+
+        # New data available! Continue with the import.
+        # Use set to remove duplicates, but store as a list for JSON support.
+        logged_sessions_by_territory[log_key] = list(set(logged_sessions_by_territory[log_key] + periods_returned))
+    else:
+        logged_sessions_by_territory[log_key] = periods_returned
+
+    # The response groups characteristic values by nomenclature,
+    # e.g. by category of jobseeker.
+    engine = db.connection_engine()
+
+    for nomenclature in data:
+        df = pd.DataFrame(serialize_table_data_from_response(nomenclature))
+
+        with Session(engine) as session:
+            stmt = pg_insert(JobSeekerStats).values(df.to_dict("records"))
+            stmt = stmt.on_conflict_do_update(
+                index_elements=JobSeekerStats.primary_key_columns(),
+                set_={key: stmt.excluded[key] for key in df.columns},
```

Comment:
De ce que je comprend de la [doc](https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#updating-using-the-excluded-insert-values) `Insert.excluded` contient toutes les colonnes donc aussi la PK j'ai l'impression, il faudrait alors l'exclure car elle on ne veux jamais la mettre à jour même si c'est avec la même valeur
```suggestion
                set_={key: stmt.excluded[key] for key in df.columns if key not in JobSeekerStats.primary_key_columns()},
```

D'ailleurs je me demande si c'est pas légèrement mieux (si ça fonctionne) de faire plutôt dans ce sens :
```suggestion
                set_={column.name: df.columns[column.name] for column in stmt.excluded if column not in JobSeekerStats.primary_key_columns()},
```